src/AMAMM.sol

```diff
@@ -2,69 +2,206 @@
-        BalanceDelta delta,
-        bytes calldata /* hookData **/
-    ) external override poolManagerOnly returns (bytes4, int128) {
-        // fee will be in the unspecified token of the swap
-        bool specifiedTokenIs0 = (params.amountSpecified < 0 ==
-            params.zeroForOne);
-        (Currency feeCurrency, int128 swapAmount) = (specifiedTokenIs0)
-            ? (key.currency1, delta.amount1())
-            : (key.currency0, delta.amount0());
-        // if fee is on output, get the absolute output amount
-        if (swapAmount < 0) swapAmount = -swapAmount;
-
-        uint256 feeAmount = (uint128(swapAmount) * SWAP_FEE_BIPS) / TOTAL_BIPS;
-        // TODO: change address(this) to the ePoche manager address
-        poolManager.take(feeCurrency, address(this), feeAmount);
-
-        return (IHooks.afterSwap.selector, feeAmount.toInt128());
+    /// -----------------------------------------------------------------------
+    /// Storage variables
+    /// -----------------------------------------------------------------------
+
+    mapping(PoolId id => uint40) internal _lastUpdatedEpoch;
+    mapping(address deposits => uint256) public _userBalance;
+    mapping(Currency currency => uint256) internal _totalFees;
+    mapping(PoolId id => mapping(uint40 => Bid)) public poolEpochManager;
+    mapping(address manager => mapping(Currency currency => uint256)) internal _fees;
+    mapping(PoolId id => address) internal _bidToken;
+
+    /// -----------------------------------------------------------------------
+    /// Getter actions
+    /// -----------------------------------------------------------------------
+
+    function getManager(PoolId id, uint40 epoch) public view returns (Bid memory) {
+        return poolEpochManager[id][epoch];
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Bidder actions
+    /// -----------------------------------------------------------------------
+    /// @inheritdoc IAmAmm
+    function bid(PoolId id, bytes7 payload, uint128 rent, uint40 _epoch) external virtual override {
+        /// -----------------------------------------------------------------------
+        /// Validation
+        /// -----------------------------------------------------------------------
+
+        address msgSender = LibMulticaller.senderOrSigner();
+
+        uint128 deposit = depositToken(id, msgSender, _epoch);
+
+        if (
+            _epoch > _getEpoch(id, block.timestamp)
+                || rent <= poolEpochManager[id][_epoch].rent.mulWad(MIN_BID_MULTIPLIER(id)) || deposit < rent * K(id)
+                || deposit % rent != 0 || !_payloadIsValid(id, payload)
+        ) {
+            revert AmAmm__InvalidBid();
+        }
+
+        uint256 prevWinner = _getDeposit(id, _epoch);
+
+        if (
+            prevWinner == 0
+                && (
+                    _lastUpdatedEpoch[id] > _getEpoch(id, block.timestamp)
```

Comment:
```suggestion
                    
```
Removing as `_lastUpdatedEpoch[id] <= _getEpoch(id, block.timestamp)` is still feasible. E.g. last is 2, k is 5, current is 3, you could still able to bid for epoch 4.